File: fsharp/nth-prime/NthPrime.fs
module NthPrime

let isMultipleOf x y = x % y = 0

let prime nth : int option =
    if nth < 1 then
        None
    else
        let mutable primes = []
        let store prime =
            primes <- primes @ [prime]
            prime

        let isPrime x = primes |> List.exists (isMultipleOf x) |> not

        Seq.initInfinite id
        |> Seq.skip 2 // 0 and 1 are not considered primes
        |> Seq.filter isPrime
        |> Seq.map store
        |> Seq.skip (nth - 1)
        |> Seq.tryHead

File: scala/nth-prime/src/main/scala/NthPrime.scala
object NthPrime {
  def prime(n: Int): Option[Int] = n match {
    case 0 => None
    case _ => primes.drop(n - 1).headOption
  }

  private def primes: Seq[Int] =
    Stream.from(2).filter(isPrime)

  private def isPrime(n: Int) =
    (2 to math.sqrt(n).toInt).forall(n % _ != 0)
}